### ISSUE TITLE
Incorrect assessment of exhausting matching for tuple with literal values

### ISSUE THREAD

**Issue by thisisrandy** [microsoft/pyright#3153]
**Describe the bug**
Variations on tuple matching arms can be constructed to report both false positive inexhaustive and false negative exhaustive matches:

```python
# pyright reports:
# Cases within match statement do not exhaustively handle all values
#  Unhandled type: "tuple[Literal[0]]"
match (0,):
    # however, this arm matches
    case (0,):
        print(1)

# equivalent to the above, to the best of my understanding, but no error
match (0,):
    case tuple((0,)):
        print(2)

# however, we can also construct a false negative
match (0,):
    # this arm does not match, but no error is reported
    case tuple((1,)):
        print(3)

match (0,):
    # pyright catches the problem in this construction:
    # Condition will always evaluate to False since the types "Literal[0]" and
    # "Literal[1]" have no overlap
    case tuple((foo,)) if foo == 1:
        print(4)

# but incorrectly reports inexhaustive matching here
match (0,):
    case tuple((foo,)) if foo == 0:
        print(5)


def match_tuple(tup: tuple[int]):
    # the tuple((literal,)) construction is also allowed when we have no
    # information about the value being matched
    match tup:
        case tuple((1,)):
            print(6)


match_tuple((0,))
match_tuple((1,))

# out:
# 1
# 2
# 5
# 6
```

**To Reproduce**
Paste the above into VS Code and observe which constructions are accepted and rejected.

**Expected behavior**
Cases 1, 2, and 5 should be accepted, and 3, 4, and 6 should be rejected (and for the right reason).

**VS Code extension or command-line**
pylance: v2022.3.0
pyright: 1.1.226

**Comment by erictraut** [comment]
This is by design. Match exhaustion makes use of type narrowing, and type narrowing applies only to simple identifiers, member access expressions of the form `a.b` and index expressions of the form `a[0]`. It does not apply to arbitrary expressions like tuples.

You have a few choices here:
1. You can choose not to use the "reportMatchNotExhaustive" diagnostic rule if you don't like how it works.
2. You can avoid using tuple expressions as a match subject.
3. You can add a catch-all case (suck as a wildcard) to the end of the match statement.

If you're seeing different behaviors between pyright and pylance, that's because you have them configured differently.

**Comment by erictraut** [comment]
Actually, on closer inspection, there might be an issue here. I'll take a closer look.

**Comment by thisisrandy** [comment]
> If you're seeing different behaviors between pyright and pylance, that's because you have them configured differently.

Yes, sorry, I had forgotten to explicitly set my mode to strict for the CLI tool. Add 

```toml
[tool.pyright]
typeCheckingMode = "strict"
```

makes reporting consistent. I'll edit the bug report.

**Comment by erictraut** [comment]
OK, here's what's happening here.

First, it's important to note that the pattern `(0,)` is completely different from the pattern `tuple((0,))`. The former is a sequence pattern, and the latter is a class pattern. They act very differently and follow very different code paths both in the runtime and in the type checker.

Second, the point I made about type narrowing was misleading (actually irrelevant to this case) because you're not using the subject expression multiple times and expecting it to be narrowed on subsequent times. So ignore that statement.

Third, the root cause here is that pyright currently never performs negative type narrowing for sequence patterns, so there's no way for the subject type to be narrowed to `Never` by the time it gets to the bottom of the match statement. The reason I didn't implement negative type narrowing for sequence patterns is that there are _very_ few cases where it's possible to do negative type narrowing. You have effectively identified the only special case where it's possible — namely, where the subject type is a tuple with a statically-known length, with known entry types that are literal values, and with sequences that use only literal patterns.

I can add support for this very specific case.

**Comment by thisisrandy** [comment]
> First, it's important to note that the pattern `(0,)` is completely different from the pattern `tuple((0,))`. The former is a sequence pattern, and the latter is a class pattern. They act very differently and follow very different code paths both in the runtime and in the type checker.

My point was that both patterns match `(0,)` and nothing else, but point taken. If I'm reading your response correctly, you're planning to address the specific sequence case, but not, to use the language of the grammar in [PEP 634](https://www.python.org/dev/peps/pep-0634/#class-patterns), the class pattern with positional patterns case, correct?

I feel like maybe I should rephrase my expectation here. This, as you put it, very specific case is contrived and grew out of some what if experimentation after reading [PEP 586](https://www.python.org/dev/peps/pep-0586/) on literal types. I'm not so concerned about false positives, because, as you said, it's easy to stick in a wildcard arm at the end. The false negatives are more concerning, however. I don't yet understand why e.g. my case 3 (reproduced below) is accepted and why we should view that as correct behavior.

```python
match (0,):
    # this arm does not match, but no error is reported
    case tuple((1,)):
        print(3)
```

If pyright can't reason about the pattern arguments portion of a class pattern but also can at least detect that a pattern argument which imposes some additional constraints exists, shouldn't it at least punt and report a non-exhaustive match?

**Comment by erictraut** [comment]
You are correct. In addition to the lack of negative type narrowing for sequence patterns, there was also a bug that affected negative type narrowing for class patterns when the class was one of the "special built in classes" designated in PEP 634 (which includes `tuple`) and a class pattern argument was provided. I've addressed both of those issues.

The other case that involves a guard expression (`if foo == 0`) is not a bug. Guard expressions are not taken into consideration for purposes of type narrowing. If a guard expression is present, the pattern is always considered conditional, and negative type narrowing is skipped. Guard expressions that can be statically determined to be True or False are probably a programming error and will be flagged as such if `reportUnnecessaryComparison` is enabled.

With these fixes in place the following errors are now reported:

```python
# pyright: strict

match (0,): # No Error
    case (0,):
        pass

match (0,):  # No Error
    case tuple((0,)):
        pass

match (0,): # Error: not exhaustive
    case tuple((1,)):
        pass

match (0,): # Error: not exhaustive
    case tuple((foo,)) if foo == 1: # Error: unnecessary comparison
        pass

match (0,): # Error: not exhaustive
    case tuple((foo,)) if foo == 0:
        pass

def match_tuple(tup: tuple[int]):
    match tup: # Error: not exhaustive
        case tuple((1,)):
            pass
```

**Comment by thisisrandy** [comment]
> The other case that involves a guard expression (`if foo == 0`) is not a bug. Guard expressions are not taken into consideration for purposes of type narrowing. If a guard expression is present, the pattern is always considered conditional, and negative type narrowing is skipped. Guard expressions that can be statically determined to be True or False are probably a programming error and will be flagged as such if `reportUnnecessaryComparison` is enabled.

Right, Rust (and probably many others) does the same thing, so I should have figured.

This all looks great. I look forward to the next version.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.228, which I just published. It will also be included in the next release of pylance.
